os/include/tinyara/preference.h

```diff
@@ -70,21 +70,17 @@ enum preference_result_error_e {
 #define PREFERENCE_CBMQ_LEN         16
 #define PREFERNENCE_CBMSG_MAX       32
 
-struct value_data_s {
+struct value_attr_s {
 	int type;
 	int len;
-	union {
-		int i;
-		double d;
-		char *s;
-	};
 };
-typedef struct value_data_s value_data_t;
+typedef struct value_attr_s value_attr_t;
 
 struct preference_data_s {
 	char *key;
 	int type;
-	value_data_t value;
+	value_attr_t attr;
+	void *value;
```

Comment:
If `preference_data_t` has `void *value` and `value` is allocated during `preference_read_fs_key`,
user have to free `value` to avoid memory leak.

User will free `preference_data_t`, but will they free `value`?
I suggest that preference supports destroy function which frees `value` and `preference_data_t`.
How do you think?

Reply:
Result values are freed except string after assigning them to user variables.
For example, private_preference.c:173
  *value = *(double *)data.value;
  PREFERENCE_FREE(data.value);
So user need to free string value only.

User just calls preference APIs with key, value data and they are packed as preference_data_t and processed internally.
It means user don't use preference_data_t directly so free functions is not needed.